File: packages/ios/Sources/Snapfill/Snapfill.swift
import Foundation
import WebKit

/// Snapfill helper — attach to any WKWebView to enable form detection,
/// cart detection, value capture, and form filling.
///
/// Usage:
/// ```swift
/// let snapfill = Snapfill(webView: webView)
/// snapfill.delegate = self
/// snapfill.attach()
/// ```
public class Snapfill: NSObject, WKScriptMessageHandler {

    public weak var delegate: SnapfillDelegate?

    private weak var webView: WKWebView?
    private let options: SnapfillOptions
    private var attached = false

    private static let messageHandlerName = "snapfill"

    private static let bridgeShim = """
    window.ReactNativeWebView={postMessage:function(m){webkit.messageHandlers.snapfill.postMessage(m);}};
    """

    private lazy var detectionScript: String = {
        Self.loadResource("snapfill", ext: "js") ?? ""
    }()

    private lazy var fillTemplate: String = {
        Self.loadResource("snapfill-fill", ext: "js") ?? ""
    }()

    public init(webView: WKWebView, options: SnapfillOptions = SnapfillOptions()) {
        self.webView = webView
        self.options = options
        super.init()
    }

    /// Attaches Snapfill to the WKWebView. Adds user scripts and a message handler.
    public func attach() {
        guard let webView = webView, !attached else { return }
        attached = true

        let controller = webView.configuration.userContentController

        // Bridge shim — inject at document start so it's available before any scripts run
        let bridgeUserScript = WKUserScript(
            source: Self.bridgeShim,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: true
        )
        controller.addUserScript(bridgeUserScript)

        // Detection scripts — inject at document end
        if options.detectForms || options.captureValues {
            let detectUserScript = WKUserScript(
                source: detectionScript,
                injectionTime: .atDocumentEnd,
                forMainFrameOnly: true
            )
            controller.addUserScript(detectUserScript)
        }

        // Message handler
        controller.add(self, name: Self.messageHandlerName)
    }

    /// Fills form fields in the current page using the provided field-to-value mappings.
    public func fillForm(_ mappings: [String: String]) {
        guard let webView = webView else { return }
        guard let jsonData = try? JSONSerialization.data(withJSONObject: mappings),
              let json = String(data: jsonData, encoding: .utf8) else { return }
        let script = fillTemplate.replacingOccurrences(of: "__SNAPFILL_MAPPINGS__", with: json)
        webView.evaluateJavaScript(script, completionHandler: nil)
    }

    /// Re-injects detection scripts into the current page.
    /// Useful after single-page navigation that doesn't trigger a full page load.
    public func reinject() {
        guard let webView = webView else { return }
        webView.evaluateJavaScript(Self.bridgeShim, completionHandler: nil)
        if options.detectForms || options.captureValues {
            webView.evaluateJavaScript(detectionScript, completionHandler: nil)
        }
    }

    /// Detaches Snapfill from the WKWebView, removing scripts and the message handler.
    public func detach() {
        guard let webView = webView, attached else { return }
        attached = false

        let controller = webView.configuration.userContentController
        controller.removeAllUserScripts()
        controller.removeScriptMessageHandler(forName: Self.messageHandlerName)
    }

    // MARK: - WKScriptMessageHandler

    public func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard message.name == Self.messageHandlerName else { return }
        guard let body = message.body as? String else { return }

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            Self.dispatchMessage(body, snapfill: self, delegate: self.delegate)
        }
    }

    // MARK: - Message parsing (internal for testing)

    static func dispatchMessage(_ msg: String, snapfill: Snapfill, delegate: SnapfillDelegate?) {
        guard let delegate = delegate else { return }
        guard let data = msg.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else { return }

        switch type {
        case "formDetected":
            if let fields = json["fields"] as? [String] {
                delegate.snapfillDidDetectFields(snapfill, fields: fields)
            }
        case "cartDetected":
            if let cartDict = json["cart"] as? [String: Any],
               let cart = parseCart(cartDict) {
                delegate.snapfillDidDetectCart(snapfill, cart: cart)
            }
        case "valuesCaptured":
            if let mappings = json["mappings"] as? [String: String] {
                delegate.snapfillDidCaptureValues(snapfill, mappings: mappings)
            }
        case "formFillComplete":
            if let resultDict = json["result"] as? [String: Any],
               let result = parseFillResult(resultDict) {
                delegate.snapfillDidCompleteFill(snapfill, result: result)
            }
        default:
            break
        }
    }

    static func parseCart(_ dict: [String: Any]) -> SnapfillCart? {
        let total = dict["total"] as? Int ?? 0
        let currency = dict["currency"] as? String
        var products: [SnapfillCartProduct] = []

        if let productDicts = dict["products"] as? [[String: Any]] {
            for p in productDicts {
                products.append(SnapfillCartProduct(
                    name: p["name"] as? String,
                    quantity: p["quantity"] as? Int ?? 1,
                    itemPrice: p["itemPrice"] as? Int ?? 0,
                    lineTotal: p["lineTotal"] as? Int ?? 0,
                    url: p["url"] as? String,
                    imageUrl: p["imageUrl"] as? String
                ))
            }
        }

        return SnapfillCart(total: total, currency: currency, products: products)
    }

    static func parseFillResult(_ dict: [String: Any]) -> SnapfillFillResult? {
        let filled = dict["filled"] as? Int ?? 0
        let total = dict["total"] as? Int ?? 0
        let failed = dict["failed"] as? [String] ?? []
        return SnapfillFillResult(filled: filled, total: total, failed: failed)
    }

    static func loadResource(_ name: String, ext: String) -> String? {
        guard let url = Bundle.module.url(forResource: name, withExtension: ext) else { return nil }
        return try? String(contentsOf: url, encoding: .utf8)
    }
}

File: packages/android/src/main/kotlin/com/snapfill/Snapfill.kt
package com.snapfill

import android.annotation.SuppressLint
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.WebView
import android.webkit.WebViewClient
import org.json.JSONArray
import org.json.JSONObject

/**
 * Snapfill helper — attach to any Android WebView to enable form detection,
 * cart detection, value capture, and form filling.
 *
 * Usage:
 * ```kotlin
 * val snapfill = Snapfill(webView)
 * snapfill.listener = object : SnapfillListener {
 *     override fun onFormDetected(fields: List<String>) { ... }
 * }
 * snapfill.attach()
 * ```
 */
class Snapfill(
    private val webView: WebView,
    private val options: SnapfillOptions = SnapfillOptions()
) {
    var listener: SnapfillListener? = null

    private val mainHandler = Handler(Looper.getMainLooper())
    private var attached = false
    private var originalClient: WebViewClient? = null

    private val bridgeShim: String by lazy {
        "window.ReactNativeWebView={postMessage:function(m){SnapfillBridge.onMessage(m);}};"
    }

    private val detectionScript: String by lazy {
        loadAsset("snapfill.js")
    }

    private val fillTemplate: String by lazy {
        loadAsset("snapfill-fill.js")
    }

    /**
     * Attaches Snapfill to the WebView. Sets up the JavaScript bridge and
     * installs a WebViewClient that injects scripts on page load.
     */
    @SuppressLint("SetJavaScriptEnabled", "AddJavascriptInterface")
    fun attach() {
        if (attached) return
        attached = true

        webView.settings.javaScriptEnabled = true
        webView.addJavascriptInterface(Bridge(), "SnapfillBridge")

        val wrappedClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView?, url: String?, favicon: android.graphics.Bitmap?) {
                super.onPageStarted(view, url, favicon)
                view?.evaluateJavascript(bridgeShim, null)
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                injectDetectionScripts(view)
            }
        }

        webView.webViewClient = wrappedClient
    }

    /**
     * Fills form fields in the current page using the provided field-to-value mappings.
     */
    fun fillForm(mappings: Map<String, String>) {
        val json = JSONObject(mappings).toString()
        val script = fillTemplate.replace("__SNAPFILL_MAPPINGS__", json)
        webView.evaluateJavascript(script, null)
    }

    /**
     * Re-injects detection scripts into the current page.
     * Useful after single-page navigation that doesn't trigger a full page load.
     */
    fun reinject() {
        webView.evaluateJavascript(bridgeShim, null)
        injectDetectionScripts(webView)
    }

    /**
     * Detaches Snapfill from the WebView, removing the JavaScript bridge.
     */
    fun detach() {
        if (!attached) return
        attached = false
        webView.removeJavascriptInterface("SnapfillBridge")
    }

    private fun injectDetectionScripts(view: WebView?) {
        if (view == null) return
        val scripts = buildString {
            if (options.detectForms || options.captureValues) {
                append(detectionScript)
            }
        }
        if (scripts.isNotEmpty()) {
            view.evaluateJavascript(scripts, null)
        }
    }

    private fun loadAsset(filename: String): String {
        return webView.context.assets.open(filename).bufferedReader().use { it.readText() }
    }

    private inner class Bridge {
        @JavascriptInterface
        fun onMessage(msg: String) {
            val listener = this@Snapfill.listener ?: return
            mainHandler.post {
                dispatchMessage(msg, listener)
            }
        }
    }

    internal companion object {
        fun dispatchMessage(msg: String, listener: SnapfillListener) {
            try {
                val json = JSONObject(msg)
                when (json.getString("type")) {
                    "formDetected" -> {
                        val arr = json.getJSONArray("fields")
                        val fields = (0 until arr.length()).map { arr.getString(it) }
                        listener.onFormDetected(fields)
                    }
                    "cartDetected" -> {
                        val cart = parseCart(json.getJSONObject("cart"))
                        listener.onCartDetected(cart)
                    }
                    "valuesCaptured" -> {
                        val mappings = parseStringMap(json.getJSONObject("mappings"))
                        listener.onValuesCaptured(mappings)
                    }
                    "formFillComplete" -> {
                        val result = parseFillResult(json.getJSONObject("result"))
                        listener.onFormFillComplete(result)
                    }
                }
            } catch (_: Exception) {
                // Ignore malformed messages
            }
        }

        fun parseCart(obj: JSONObject): SnapfillCart {
            val products = mutableListOf<SnapfillCartProduct>()
            val arr = obj.optJSONArray("products") ?: JSONArray()
            for (i in 0 until arr.length()) {
                val p = arr.getJSONObject(i)
                products.add(
                    SnapfillCartProduct(
                        name = p.optString("name", null),
                        quantity = p.optInt("quantity", 1),
                        itemPrice = p.optInt("itemPrice", 0),
                        lineTotal = p.optInt("lineTotal", 0),
                        url = p.optString("url", null),
                        imageUrl = p.optString("imageUrl", null)
                    )
                )
            }
            return SnapfillCart(
                total = obj.optInt("total", 0),
                currency = obj.optString("currency", null),
                products = products
            )
        }

        fun parseFillResult(obj: JSONObject): SnapfillFillResult {
            val failed = mutableListOf<String>()
            val arr = obj.optJSONArray("failed") ?: JSONArray()
            for (i in 0 until arr.length()) {
                failed.add(arr.getString(i))
            }
            return SnapfillFillResult(
                filled = obj.optInt("filled", 0),
                total = obj.optInt("total", 0),
                failed = failed
            )
        }

        fun parseStringMap(obj: JSONObject): Map<String, String> {
            val map = mutableMapOf<String, String>()
            obj.keys().forEach { key ->
                map[key] = obj.getString(key)
            }
            return map
        }
    }
}
